src/ripple/nodestore/ShardValidation.md

```diff
@@ -0,0 +1,130 @@
+# Downloaded Shard Validation
+
+## Overview
+
+In order to validate shards that have been downloaded from file servers (as opposed to shards acquired from peers), the application must confirm the validity of the downloaded shard's last ledger. The following sections describe this confirmation process in greater detail.
+
+## Execution Concept
+
+### Flag Ledger
+
+Since the number of ledgers contained in each shard is always a multiple of 256, a shard's last ledger is always a flag ledger. Conveniently, the application provides a mechanism for retrieving the hash for a given flag ledger:
+
+```C++
+boost::optional<uint256>
+hashOfSeq (ReadView const& ledger,
+           LedgerIndex seq,
+           beast::Journal journal)
+```
+
+When validating downloaded shards, we use this function to retrieve the hash of the shard's last ledger. If the function returns a hash that matches the hash stored in the shard, validation of the shard can proceed.
+
+### Caveats
+
+#### Later Ledger
+
+The `getHashBySeq` function will provide the hash of a flag ledger only if the application has stored a later ledger. When validating a downloaded shard, if there is no later ledger stored, validation of the shard will be deferred until a later ledger has been stored.
+
+We employ a simple heuristic for determining whether the application has stored a ledger later than the last ledger of the downloaded shard:
+
+```C++
+// We use the presence (or absense) of the validated
+// ledger as a heuristic for determining whether or
+// not we have stored a ledger that comes after the
+// last ledger in this shard. A later ledger must
+// be present in order to reliably retrieve the hash
+// of the shard's last ledger.
+if (app_.getLedgerMaster().getValidatedLedger())
+{
+    auto const hash = app_.getLedgerMaster().getHashBySeq(
+        lastLedgerSeq(shardIndex));
+
+    .
+    .
+    .
+}
+```
+
+The `getHashBySeq` function will be invoked only when a call to `LedgerMaster::getValidatedLedger` returns a validated ledger, rather than a `nullptr`. Otherwise validation of the shard will be deferred.
```

Comment:
> On further reflection, maybe `LedgerMaster::getValidatedLedger()` really is the call you want to make, but its not for the reason you explained. A non-null return from `LedgerMaster::getValidatedLedger()` _could_ be used as a proxy for whether the server has at any point in the past been synced to the network. Is that really the condition you're checking for?...

Your last reflection is correct. We're using the non-null return from `LedgerMaster::getValidatedLedger()` as an indication that the node has been synched to the network.

Reply:
Okay, but is that really the most useful check?  I think it just says that the server has at some point in the past been synchronized with the network.  I don't think it says that the server is _currently_ synchronized with the network.  The server may have been in sync at some point in the past, but fallen out of sync three days ago.  Is that still a useful condition to check?  If so, please explain why in a comment.

Reply:
@scottschurr  Great questions! 

The ledger returned from `getValidatedLedger()` will always have a sequence larger than the flag ledger we are looking for, unless the shard was created on a different network. In that case we will correctly end up failing validation and deleting the shard. Any validated ledger with a sequence larger than the flag ledger we are looking for will provide the hash we want.

There isn't a need to be in sync with the network but calling `getValidatedLedger()` is a simple solution. It has the benefit of deferring the storage IO necessary for validation while the server is trying to sync.

The alternative is to query the main SQLite database for the flag ledger hash we are looking for. Unless the server holds a lot of history it will likely fail to find it. Since every ledger has the skip list that includes the hash of every preceding flag ledger, the fallback would then be to query for a ledger hash with sequence larger than the flag ledger we are looking for. If found, we would use that hash to load the ledger and its skip list from the node store. If the server doesn't have suitable history stored, we would have to defer until we acquire a validated ledger.

Although the alternate solution might be faster at obtaining the hash we are looking, I don't know if it warrants the additional complexity.

Reply:
@miguelportilla thanks for the answer.  I'm still a bit puzzled...

>The ledger returned from getValidatedLedger() will always have a sequence larger than the flag ledger we are looking for, unless the shard was created on a different network.

I understand why that's _usually_ the case, but I don't think it's necessarily _always_ the case.  Suppose a server has been in sync with the network and falls out of sync shortly before a flag ledger.  The flag ledger passes and the server is still not in sync, and the user requests acquiring the most recent shard on the network.  At this point a valid shard with a more recent ledger sequence than the most recent _local_ validated ledger can be acquired.  That shard will then eventually be discarded unless the server regains sync within a reasonable time frame (about 5 minutes these days).

It seems like this could all be avoided by calling `LedgerMaster::getValidLedgerIndex()` (which returns a `uint32_t` ledger sequence), rather than calling `LedgerMaster::getValidatedLedger()`.

`LedgerMaster::getValidLedgerIndex()` has very low overhead (read an atomic `uint32_t`).  `LedgerMaster::getValidLedgerIndex()` also returns a value that you can compare against.  If the returned value is less than (or equal to?) the index of the last ledger in the shard then you can't validate the shard yet.  You must wait.

Does that distinction make sense?  It feels like I'm still missing something.  Sorry to be slow on the uptake.

Reply:
@scottschurr Good point, that could happen and I didn't consider it. As you suggested, using `LedgerMaster::getValidLedgerIndex()` and comparing its value against the shard's last ledger sequence would avoid the problem. Thanks for catching this!